JDK-8270147: Increase stride size allowing unrolling more loops

rsmogura opened the pull request:
# Description

Increase allowed stride size for loop unrolling to the maximum vector size on runtime platform. 

The motivation for this change is discussion and research about unrolling vector (SIMD) loops. For vector usage, stride size depends on vector element type, and platform vector size. For AVX256 and int stride size is 8, and loop unroll happens. However short and byte loops could not get unrolled (stride size 16 & 32):
```java
    for (int i = 0; i < SPECIES.loopBound(longSize); i += SPECIES.length() /* 8 for int, 16 for short */ ) { 
      var v = ShortVector.fromByteBuffer(SPECIES, srcBufferHeap, i, ByteOrder.nativeOrder()); 
      v.intoByteBuffer(dstBufferHeap, i, ByteOrder.nativeOrder()); 
    } 
```
After this change, the maximum stride, which allows loops to unroll, will depend on the maximum bytes size of vectors registers (AVX256 - 32, AVX512 - 64, SVE up to 256)

# Notes
Stride size was decreased some time ago https://github.com/openjdk/panama-foreign/commit/2683d5390bd58683ae13bdd8582127c308d8fd04

The exact reasons for this are not known for me (over unroll of some loops?).

Original thread https://mail.openjdk.java.net/pipermail/panama-dev/2021-June/014310.html

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270147](https://bugs.openjdk.java.net/browse/JDK-8270147): Increase stride size allowing unrolling more loops


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Igor Veresov](https://openjdk.java.net/census#iveresov) (@veresov - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4658/head:pull/4658` \
`$ git checkout pull/4658`

Update a local copy of the PR: \
`$ git checkout pull/4658` \
`$ git pull https://git.openjdk.java.net/jdk pull/4658/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4658`

View PR using the GUI difftool: \
`$ git pr show -t 4658`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4658.diff">https://git.openjdk.java.net/jdk/pull/4658.diff</a>

</details>

vnkozlov commented:
RFR for [7039652](https://bugs.openjdk.java.net/browse/JDK-7039652) changes 
https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2011-April/005314.html

As RFR said, I assumed that stride check was used to limit unrolling to 16 (when initial stride is '1'). That is how `MAX_UNROLL` was introduced.

The stride check
```
   // Check for initial stride being a small enough constant
   if (abs(cl->stride_con()) > (1<<2)*future_unroll_ct) return false;
```
as comment said is equivalent to next check regardles of unrolling state:
```
  if (initial_stride_con > 8) return false;
```

General note about unrolling. There are two main issues:
1. Over-unroll - main loop will be skipped and not optimized post-loop will execute all iterations.
2. Main loop body size become too big so that fetching of loop's instructions become expensive (my interpretation).

I don't think skipping stride check (checking only for Integer overflow) is correct solution for current issue.
May be use `Matcher::max_vector_size()` instead:
```
   // Check for initial stride being a small constant but not smaller than max vector size
   int stride_limit = Matcher::max_vector_size(T_BYTE);
   if (abs(cl->stride_con()) > (stride_limit * future_unroll_ct)/2) return false;
```

rsmogura commented:
Hi Vladimir & thank you for comments.

I really don't understand the limit `1<< 2` for initial stride size, as loop unroll count is determined by LoopMaxUnroll.

I adjusted code to your suggestions, right now `initial_stride_sz = MAX2(1<<2, Matcher::max_vector_size(T_BYTE) / 2)`

I dropped the JVM flag controlling what procedure should be used to determine max stride size, is it ok?

vnkozlov commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -2001,8 +2006,8 @@ void PhaseIdealLoop::do_unroll(IdealLoopTree *loop, Node_List &old_new, bool adj
-  assert(old_trip_count > 1 &&
-      (!adjust_min_trip || stride_p <= (1<<3)*loop_head->unrolled_count()), "sanity");
+  assert(old_trip_count > 1 && (!adjust_min_trip || stride_p <=
+    MIN2(max_jint / 2 - 2, MAX2(1<<2, Matcher::max_vector_size(T_BYTE)) * loop_head->unrolled_count())), "sanity");
```
Should be `1<<3` here. Otherwise looks good.

rsmogura replied:
Thank you, fixed.

vnkozlov commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -2001,8 +2006,8 @@ void PhaseIdealLoop::do_unroll(IdealLoopTree *loop, Node_List &old_new, bool adj
-  assert(old_trip_count > 1 &&
-      (!adjust_min_trip || stride_p <= (1<<3)*loop_head->unrolled_count()), "sanity");
+  assert(old_trip_count > 1 && (!adjust_min_trip || stride_p <=
+    MIN2(max_jint / 2 - 2, MAX2(1<<3, Matcher::max_vector_size(T_BYTE)) * loop_head->unrolled_count())), "sanity");
```
I assume you need to do the same `MIN2<int>` fix here. Try debug build where asserts are included.

rsmogura replied:
Thank you, I missed this. Debug build crashed and I didn't restart it.

vnkozlov reviewed (commented):
Good. I am testing v04.

Do you have benchmark to verify the fix (vectorization for `byte` and `short` arrays)? Consider adding it to `test/micro/org/openjdk/bench/jdk/incubator/foreign/`

You need second review.

rsmogura commented:
Thank you Vladimir.

Benchmarks add.

> Good. I am testing v04.
> 
> Do you have benchmark to verify the fix (vectorization for `byte` and `short` arrays)? Consider adding it to `test/micro/org/openjdk/bench/jdk/incubator/foreign/`
> 
> You need second review.

I pushed benchmarks with a new commit (I can't take a lot of credit for benchmarks), and as side note the ByteBuffers improvements are in another question / PR.

I hope this change will not unroll loops that should not get unrolled.

Thank you review, and sorry for the issues with Win build.

vnkozlov reviewed (changes requested):
Thank you for adding benchmarks but, please, add copyright header to them (see examples in other files in directory).

vnkozlov commented:
Testing tier1-3 passed. Good.

vnkozlov reviewed (approved):
Good.
I submitted SPEC benchmarks testing.

rsmogura commented:
Hi Vladimir, and thank you for the review.

You mentioned spec tests I wonder how did it go? Is it 'safe'?

vnkozlov commented:
Perf tests passed - I don't have concerns. It took time.

You still need second review.

veresov reviewed (approved):
Looks reasonable.

rsmogura commented:
/integrate

kelthuzadx commented:
/sponsor
